lib/pbench/test/functional/agent/cli/commands/tools/test_clear_tools.py

```diff
@@ -309,37 +308,51 @@ def test_clear_tools_test85(pbench_run, tools_configuration, groups, remotes, to
 
 
 @pytest.mark.parametrize(
-    "groups", ["default", "default,wrong_group", "group1,wrong_group"],
+    "groups", ["", "default,wrong_group", "group1,wrong_group"],
 )
 @pytest.mark.parametrize(
     "remotes", ["", "remote1,wrong_remote1", "wrong_remote1,wrong_remote2"],
 )
 @pytest.mark.parametrize(
-    "tools", ["", "pidstat,mpstat,wrong_tool1", "wrong_tool1,wrong_tool2"],
+    "tools", ["", "pidstat,wrong_tool1", "wrong_tool1,wrong_tool2"],
 )
 def test_clear_tools_test86(tools_configuration, groups, remotes, tools):
     """ Attempt to clear tools at various group-remote-tool locations where the
     combination may not present on the hosts"""
-    command = [
-        "pbench-clear-tools",
-        f"--groups={groups}" if groups else "",
-        f"--remotes={remotes}" if remotes else "",
-        f"--name={tools}" if tools else "",
-    ]
-    command_refined = [x for x in command if x]
+    _, tools_tree = tools_configuration
+    command = ["pbench-clear-tools"]
+    if groups:
+        command.append(f"--groups={groups}")
+    if remotes:
+        command.append(
+            f"--remotes={','.join(map('{}.example.com'.format, remotes.split(',')))}"
+        )
+    if tools:
+        command.append(f"--name={tools}")
 
-    out, err, exitcode = pytest.helpers.capture(command_refined)
+    out, err, exitcode = pytest.helpers.capture(command)
     assert b"" == out
-    if groups != "default":
-        assert any(
-            f'pbench-clear-tools: invalid --group option "{x}"'.encode() in err
-            for x in groups.split(",")
-        )
-    if remotes:
-        assert any(f'No remote host "{x}"'.encode() in err for x in remotes.split(","))
-    if (
-        tools == "wrong_tool1,wrong_tool2"
-        and "wrong_remote" not in remotes
-        and "wrong_group" not in groups
-    ):
-        assert b"Tools ['wrong_tool1', 'wrong_tool2'] not found in remote" in err
+
+    if not groups:
+        groups = "default"
+
+    for group in groups.split(","):
+        err_msg = f'pbench-clear-tools: invalid --group option "{group}"'
+        if group not in tools_tree.keys():
+            assert err_msg.encode() in err
+        else:
+            assert err_msg.encode() not in err
+            for remote in remotes.split(","):
+                err_msg = f'No remote host "{remote}.example.com"'
+                if remote and remote not in tools_tree[group]:
+                    assert err_msg.encode() in err
+                elif remote:
+                    wrong_tools = []
```

Comment:
> First, for the permutations set at line 314, `remote` will always be `True`, so testing it at lines 347 and 349 doesn't accomplish anything. But, it does raise the issue that we need code (and testing) for values passed to `--groups`, `--remotes`, and `--name` which either start or end with a comma! These will result in empty values for their respective items (e.g., in this case the empty `group` will _not_ be translated to `"default"`), and we'll run into lots of problems referencing directories and files with null names).

Ah yeah that is an interesting point, someone can provide options like `--groups=group1,`, I dont know how likely it is but by adding a check for each value in the list against an empty string (in this case `["group1", ""]`) we could handle that situation.

> Second, with the code as it stands presently, when `remotes` is empty, we're going to skip all the remotes and tools assertions. However, now that we have access to the `tools_tree`, we should be able to generate the appropriate default value for `remotes` (i.e., all of the ones under the `group`) to replace the empty string. Likewise, for `tools`. This should enable the rest of the code here to work.

yeah I guess but its bit more complicated, if remotes is empty we send messages similar to the following on stderr pipe:
```
(b'Removed "pidstat" from host "remote1.example.com" in tools group "defaul'
b't"\nAll tools removed from group "default" on host "remote1.example.com"\n'
b'Removed "mpstat" from host "remote2.example.com" in tools group "default'
b'"\nAll tools removed from group "default" on host "remote2.example.com"\nR'
b'emoved "pidstat" from host "remote3.example.com" in tools group "default'
b'"\nRemoved "turbostat" from host "remote3.example.com" in tools group "de'
b'fault"\nAll tools removed from group "default" on host "remote3.example.c'
b'om"\n')
```
so we could make an assertion on whether all the remotes and tools under that group are present in our message (which it should), is that what you mean?

Reply:
> Ah yeah that is an interesting point, someone can provide options like --groups=group1,

Indeed; or `--groups=,a,,b,,,,,,,c,,` ... failing after the `split` result if there are any empty terms might be good. Too bad `click` doesn't support this directly. (Their "standard" alternative would be to support repetition of the `--group` option to specify multiple groups, like `--group a --group b`.)

Reply:
With the latest revision, the handling of `remotes` here still isn't quite right:  there are two cases to be addressed -- a completely blank `remotes` value, and a single blank `remote` in the list.  I think this code does the right thing (more or less) if `remotes` is completely blank; however, I think it also does that same thing for each blank `remote` in the list...and I don't know whether that is merely unsatisfying or outright incorrect.

You've added special handling deep inside the loop for the "blank case".  I suggest that you hoist that out of the loop and execute it only if all of `remotes` is blank; otherwise, enter the loop and process each value, handling the blank ones appropriately (presumably by skipping them, unless I've changed your mind to call them out as errors).

Reply:
> I think this code does the right thing (more or less) if `remotes` is completely blank; however, I think it also does that same thing for each blank `remote` in the list

Well, for this test (test86) there are no values like this `remote1,,,remote1` in the parametrization so the list is either empty or the format is correct. 

But as the code stands right now if someone provides value like this `remote1,,,remote1` it will be translated to this `[remote1, remote2]` and so there is nothing to check against when the value is empty, we will just have to skip them so that's why I didnt use them in this test case parameterization

Reply:
> Well, for this test (test86) there are no values like this `remote1,,,remote1` in the parametrization so the list is either empty or the format is correct.

Yeah, I know, but counting on that makes the code kind of fragile, since the test code can then handle only very specific inputs.

> But as the code stands right now if someone provides value like this `remote1,,,remote1` it will be translated to this `[remote1, remote2]` and so there is nothing to check against when the value is empty, we will just have to skip them so that's why I didnt use them in this test case parameterization

The translation you're referring to happens in the CUT, but the test code still iterates over all of the items, including any blank ones.  So, necessarily, the test behavior should be different depending on whether there are blank options specified or _no_ options specified (and it's not...).  As you point out, we wouldn't otherwise notice, because this test is never run with individual blank options, but I suspect that it could and perhaps _should_ be.  Having the handling for no-options-specified inside the loop which handles individual options is kind of weird (e.g., it depends on `"".split(",")` returning exactly one, empty-string item...which is possibly a bit of a stretch), and moving it to a more appropriate position would free up that slot for handling individual blank options (however we decide that they are supposed to be handled).